classes/Cart.php

```diff
@@ -1428,7 +1428,7 @@ public function updateQty(
 
             if ($operator == 'up') {
                 $updateQuantity = '+ ' . $quantity;
-                $newProductQuantity = $productQuantity - $quantity;
+                $newProductQuantity = ($productQuantity + $cartProductQuantity['quantity']) - $quantity;
```

Comment:
We get the product quantity from `Product::getQuantity()`, but for some reason this method subtract the number of product of the same kind in the cart from the total quantity, so we need to add it here to do our comparison.

I was thinking about replacing the use of `Product::getQuantity()` for something else, but I'm not sure about the possible consequences. This fix doesn't break any test.

Reply:
This is wanted, when the cart is provided as a parameter to `Product::getQuantity` it is meant to substract the cart's quantity from the stock Because as long as a cart is not finalized the stock is not decreased

The real stock is only substracted when the order is created (because a pending cart is not considered to be finished so it doesn't affect the stock yet)

That's why it MUST performs this computation, here's a example because it's not an easy topic:
- Product A has 5 quantities of stock
- Cart has 3 product A
- So "virtually" there is only 2 left Product A in stock
- I update the cart and add 3 more Product A (so 6 in total which is not possible)
- The action in the cart will "up" of 3 products
- So Product::getQuantity returns 5 (real stock) - 3 (products present in cart) = 2 virtually available products
- Then the action must fail because it's too high

So your modif will break this behaviour The biggest problem is we have ZERO test to ensure we don't break this.. 

Now in your case it's different because you are modifying a cart which has already been ordered, so the stock has already been updated and is accurante regardles of the cart quantity, so you can't rely on the same rules (but it doesn't mean you can modify them or it will break the FO behaviour)

We already encountered this kind of issues in other Order handlers, usually the solution is to perform all the stock checks beorehand to make sure they are valid and disbale this check using the last parameter of `Cart::updateQuantity` => `$skipAvailabilityCheckOutOfStock = true`

Like here:
https://github.com/PrestaShop/PrestaShop/blob/611916bdb01be94bbb094e000f3b6e2a99590330/src/Adapter/Order/OrderProductQuantityUpdater.php#L251

Reply:
The problem here is that the gift present is automatically added via:

https://github.com/PrestaShop/PrestaShop/blob/611916bdb01be94bbb094e000f3b6e2a99590330/classes/Cart.php#L1216

So it's hard coded to check the stock based on the cart quantity And you can't even add a new parameter in `addCartRule` that would ignore this when called from an Order handler, because it's NOT called from a handler It is automatically called when the product is added and `CartRule::autoAddToCart` is called

So that's a tricky one ^^ But I don't have the solution right now

Reply:
The whole problem is about the fact that Cart is actually ordered but we have no way to know that with the cart's data alone.. One solution could be to add an `id_order` column in the Cart that is automatically filled when the order is created

This way we could easily detect that a cart is an order cart (and it would probably allow other things later, or for some refacto) But I'm not even sure it would be enough, because I'm not sure it's safe to simply disable some checks because a cart is ordered

That said.. maybe it would allow us to update `Product::getQuantity` and it wouldn't substract the cart quantity if the cart has an order (meaning stock is already udpated) That's a lead.. But it would be safer to add behat tests on the cart to avoid too many regressions

Reply:
thank you for those very useful informations ! I'll modify `Product::getQuantity` so that it subtracts the cart quantity only if the cart has no related order. And add tests for the case you described above. That will take the addition of a new column in ps_cart.

Reply:
I tried the scenario you wrote above, it didn't seem to be broken even with my former fix...  I still made the fix we talked about. 

I think the kind of scenario you're worried about is throughly tested here: `tests/Integration/Behaviour/Features/Scenario/Order/order_out_of_stock.feature`